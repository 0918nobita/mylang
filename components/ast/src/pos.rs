use std::fmt;

use serde::{Deserialize, Serialize};

#[derive(Clone, Default, Serialize, Deserialize)]
pub struct Pos {
    line: u32,
    column: u32,
}

impl fmt::Debug for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.line, self.column)
    }
}
