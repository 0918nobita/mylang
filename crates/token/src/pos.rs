use std::fmt;

use serde::{Deserialize, Serialize};

#[derive(Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Pos {
    line: u32,
    character: u32,
}

impl Pos {
    pub fn new(line: u32, character: u32) -> Self {
        Self { line, character }
    }

    /// 次の文字位置に移動する (改行ではない文字が続くことを想定している)
    ///
    /// ```
    /// use token::Pos;
    ///
    /// let mut pos = Pos::new(3, 7);
    /// pos.next_char();
    ///
    /// let expected = Pos::new(3, 8);
    ///
    /// assert_eq!(pos, expected);
    /// ```
    pub fn next_char(&mut self) {
        self.character += 1;
    }

    /// 次の行の先頭に移動する
    ///
    /// ```
    /// use token::Pos;
    ///
    /// let mut pos = Pos::new(1, 8);
    /// pos.next_line();
    ///
    /// let expected = Pos::new(2, 0);
    /// assert_eq!(pos, expected);
    /// ```
    pub fn next_line(&mut self) {
        self.line += 1;
        self.character = 0;
    }
}

impl fmt::Debug for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.line, self.character)
    }
}

impl fmt::Display for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.line + 1, self.character + 1)
    }
}
