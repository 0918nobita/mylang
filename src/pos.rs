use std::fmt;

#[derive(Default)]
pub struct Pos {
    line: u32,
    column: u32,
}

impl fmt::Debug for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.line, self.column)
    }
}
