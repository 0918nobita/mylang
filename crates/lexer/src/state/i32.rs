use mylang_token::{range, Pos, Token};

#[derive(Clone, Debug, Default)]
pub struct I32State {
    start: Pos,
    end: Pos,
    acc: String,
}

impl I32State {
    pub fn new(pos: Pos, acc: String) -> Self {
        Self {
            start: pos.clone(),
            end: pos,
            acc,
        }
    }

    pub fn append_digit_char(&self, pos: Pos, c: char) -> Self {
        Self {
            start: self.start.clone(),
            end: pos,
            acc: format!("{}{c}", self.acc),
        }
    }

    pub fn tokenize(&self) -> Token {
        let i = self.acc.parse::<i32>().unwrap();
        Token::I32(range!(self.start.clone(), self.end.clone()), i)
    }
}

#[cfg(test)]
mod tests {
    use mylang_token::pos;

    use super::I32State;

    #[test]
    fn test_i32_state() {
        let state = I32State::default();
        let state = state.append_digit_char(pos!(0;0), '1');
        let state = state.append_digit_char(pos!(0;1), '2');
        let token = state.tokenize();
        insta::assert_debug_snapshot!((state, token));
    }
}
