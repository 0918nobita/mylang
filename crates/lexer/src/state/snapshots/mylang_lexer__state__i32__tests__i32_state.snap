---
source: crates/lexer/src/state/i32.rs
assertion_line: 45
expression: "(state, token)"

---
(
    I32State {
        start: 0:0,
        end: 0:1,
        acc: "12",
    },
    I32(
        0:0-0:1,
        12,
    ),
)
